---
title: "Dynamic eQTLs during development and heat stress"
author: Cas Blaauw
output: html_document
---

# Load libraries and data

### Libraries
```{r}
library(openxlsx)
library(gplots)
library(cowplot)
library(tidyverse)
library(reshape2)
library(VennDiagram)
library(RColorBrewer)
library(randomForest)
library(randomForestExplainer)
library(glue)
library(magrittr)
library(tictoc)
library(furrr)
```

### Expression data

Describes the expression for each gene for each worm.

```{r}
gexp <- list(
  dev = read.delim("Data/Snoek_Sterken_CT_gene_expression_RILs.txt", sep = "\t"),
  hs = read.delim("Data/Snoek_Sterken_HS_gene_expression_RILs.txt", sep = "\t"),
  rec = read.delim("Data/Snoek_Sterken_REC_gene_expression_RILs.txt", sep = "\t")
)
head(gexp$dev[,1:5])
dim(gexp$dev)
```

### Projection axis

Describes the developmental, heat shock and recovery state for each worm.

```{r}
load("Data/projection_alldata.rdata")
projection_all <- mutate(projection_all, dynamic = tolower(experimentype))
dim(projection_all)
```

### Gene map

Describes the present genomic sections/alleles in each worm.

```{r}
gm <- list(
  dev = read.delim("Data/map_CT.txt", sep = "\t"),
  hs = read.delim("Data/map_HS.txt", sep = "\t"),
  rec = read.delim("Data/map_REC.txt", sep = "\t")
)

head(gm$dev[,1:5])
```

### Genome info

Maps the genetic regions/alleles to coordinates on the genome.

```{r}
mrk <- read_tsv("Data/marker.txt", show_col_types = FALSE)
head(mrk)
dim(mrk)
```

### Transcriptome info

Maps the expressed genes to coordinates on the genome.

```{r}
gi <- read_csv("Data/WormGenes_all.csv", col_names = c("gene", "name", "start", "end", "chr"), show_col_types = FALSE)
head(gi)
dim(gi)
```

### Random forest results
```{r}
load("Data/RF_results2/data_incmse_dev.out")
load("Data/RF_results2/data_incmse_hs.out")
load("Data/RF_results2/data_incmse_rec.out")

load("Data/RF_results2/data_pur_dev.out")
load("Data/RF_results2/data_pur_hs.out")
load("Data/RF_results2/data_pur_rec.out")
```


# Prepare the data

```{r}
# Combine projection, gene map, and gene expression for each dynamic into one list of dfs
samples <- map(
  c(dev = 'dev', hs = 'hs', rec = 'rec'), 
  function(dynamic) {
    # Subset projection to only include current dynamic
    proj_df <- filter(projection_all, dynamic == dynamic, line == "RIL") %>% 
      distinct(genotype, .keep_all = TRUE) %>% # Deduplicate worms within one dynamic
      select(genotype, dynamic, contains('projection'))
    
    # Load gene map for this dynamic
    gm_df <- gm[[dynamic]] %>% 
      distinct(.keep_all = TRUE) %>% 
      t() %>% # Transpose to get worms as rows
      as.data.frame() %>% 
      rownames_to_column("genotype")
    
    # Load gene expression for this dynamic
    gexp_df <- gexp[[dynamic]] %>% 
      t() %>% # Transpose to get worms as rows
      as.data.frame() %>% 
      rownames_to_column("genotype")
    
    # Combine proj, gm and gexp into one data frame
    # Use inner_join to keep only those worms that are present in all three datasets
    merged <- inner_join(proj_df, gm_df, by = 'genotype') %>% 
      inner_join(gexp_df, by = 'genotype')
    
    return(merged)
  })
```



```{r}
dynamic_type = "dev"
gene_name = "WBGene00019969"

samples_x <- samples %>% 
  extract2(dynamic_type) %>% 
  select(genotype, glue("{dynamic_type}projection"), contains("Pred")) %>% 
  column_to_rownames("genotype")
samples_y <- samples %>% 
  extract2(dynamic_type) %>% 
  select(contains("WBGene"))
```


```{r}
plan(multisession, workers = 3) # Make map parallel to speed up random forests
tictoc::tic()
rf_results <- samples_y %>%
  select(seq(1, ncol(.), 100)) %>% #Subset to every 100 genes if desired
  future_map_dfr(
    ~ randomForest(
      .x ~ ., # Match input from map (.x, gene expression column) with all variables (.) from data (samples_x aka projection+markers)
      data = samples_x,
      proximity = TRUE,
      importance = TRUE,
      ntree = 5000,
      nodesize = 2,
      maxnodes = 8
    )$importance[, 1], 
    .id = 'gene',
    .options = future_options(seed = TRUE)
  )
tictoc::toc()
```



